CoreDataStack 클래스의 DispatchQueue(serial async)에서 동작하는 fetch 메서드 구현 및 Core data Entity 재설계

donggeonoh opened the pull request:
## 구현내용
### [feat]
- CoreData
  - POI 의 값을 read, write 하는 overhead를 줄이기 위해 엔티티 분리 및 연관관계 추가
  - CoreData MO 객체 -> VO 구조체를 반환하는 연산 프로퍼티 추가

- CoreDataStack
  - 모든 좌표 데이터를 fetch 하는 메서드 추가 (serial async로 동작)
  - 특정 범위 내 좌표 데이터를 fetch 하는 메서드 추가 (sync, serial async)

- POIService
  - 메서드 내 forEach 로직을 map으로 리팩토링
  - CoreDataStack의 추가된 기능에 대한 fetch 메서드(sync, sirial async) 추가

### 첨부 사진
  <img width="658" alt="Screen Shot 2020-12-02 at 6 24 16 PM" src="https://user-images.githubusercontent.com/18098363/100853852-981d6200-34cb-11eb-9e3f-9dbd4a37f775.png">
  
  > 사진 상에서 POIinfo의 연관관계가 보이지 않지만 실제로 연관관계 입니다.

donggeonoh committed:
[feat] Core data Entity 재설계

Core data의 POI, Coordinate, Info 객체를 따로 나누고 relation 추가

donggeonoh committed:
Merge branch 'develop' into feature/core-data-architecture

donggeonoh committed:
[feat] Core data stack 및 POIService 수정

Core data stack 클래스에서 fetch의 return을 POI -> Coordinate로 수정
POIService 에서도 Coordinate 배열을 반환하도록 수정

donggeonoh committed:
[feat] Core data stack coordinate로 predicate 함수 구현

donggeonoh committed:
[feat] Core data serial Queue 비동기 작업 추가

CoreDataStack 클래스에서 fetch 시 비동기로 동작하도록 실행
CoreDataStack 에서 특정 범위의 coordinates 를 가져오는 fetch 함수 작성
Core data MO 에서 MO -> VO 로 바꿔주는 computed property 추가

donggeonoh committed:
[feat] POIService 고차함수 리팩토링 및 fetchAsync 메서드 작성

POIService forEach로 작성된 코드를 map으로 수정하여 간결하게 리팩토링
fetchAsync 메서드 추가

donggeonoh committed:
Merge branch 'develop' into feature/core-data-architecture

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/CoreData/CoreDataStack.swift:
```diff
@@ -16,6 +16,8 @@ final class CoreDataStack: DataManagable {
     private lazy var container: NSPersistentContainer? = appDelegate?.container
     private lazy var context: NSManagedObjectContext? = container?.viewContext
     
+    private let queue: DispatchQueue = .init(label: "CoreDataStackQueue", qos: .background, attributes: .concurrent)
```
nit: string은 따로 관리할 수 있도록 하는게 어떨까요

donggeonoh replied:
수정하였습니다~

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/CoreData/CoreDataStack.swift:
```diff
@@ -27,44 +29,95 @@ final class CoreDataStack: DataManagable {
     
+    func fetchAsync(topLeft: Coordinate, bottomRight: Coordinate, handler: @escaping ([POICoordinateMO]) -> Void) {
+        queue.async { [weak self] in
+            guard let self = self else { return }
+            let fetchRequest: NSFetchRequest<NSFetchRequestResult> = POICoordinateMO.fetchRequest()
+            fetchRequest.predicate = NSPredicate(format: "lng >= %@ AND lng <= %@ AND lat >= %@ AND lat <= %@",
+                                                 topLeft.x, bottomRight.x, bottomRight.y, topLeft.y)
+            guard let context = self.context,
+                  let entities = try? context.fetch(fetchRequest) as? [POICoordinateMO]
```
nit: guard를 하나로 통일 시킬수 있을거같아요

donggeonoh replied:
수정하였습니다!

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/CoreData/CoreDataStack.swift:
```diff
@@ -27,44 +29,95 @@ final class CoreDataStack: DataManagable {
         do {
             try container.persistentStoreCoordinator.execute(deleteRequest, with: container.viewContext)
         } catch let error as NSError {
             print(error.localizedDescription)
```
error를 print말고 다른걸로 처리해줘야할거같은데, 한번 같이 의논해봐요

donggeonoh replied:
넵 다음에 의논해봐요!

donggeonoh committed:
[chore] project.pbxproj 환경파일 수정

donggeonoh committed:
[fix] CoreDataStack 클래스 상수 분리 및 메서드 옵셔널 바인딩 수정

beggu84 commented on Project17-C-Map/InteractiveClusteringMap/CoreData/CoreDataStack.swift:
```diff
@@ -10,61 +10,113 @@ import CoreData
         }
+        poiMO.setValues(id: poi.id, coordinate: coordMO, info: infoMO)
+        coordMO.setValues(Coordinate(x: poi.x, y: poi.y))
+        infoMO.setValues(POIInfo(name: poi.name, imageUrl: poi.imageUrl, category: poi.category))
```
얘네 setValues 라는 메서드가 없다고, 오류가 나네요?

donggeonoh replied:
앗 제가 그부분 코드를 안올렸나보네요
수정하겠습니다!!

beggu84 reviewed (approved):
`CoreDataStack` 의 두 `fetchAsync` 안 내용이 비슷해서 하나의 함수로 뽑을 수 있어 보이네요.

`POIInfo`, `POIMO`, `PoiInfoMO` .. 헤깔리네요ㅎㅎ
MO 라고 줄여서 붙이는 거는 일반적인 컨벤션인가요? 아니면 취향인가요?

donggeonoh committed:
[feat] Entity object 클래스에 setValues 메서드 추가

poi, coordinate, info 에 대한 managed object 클래스에 setValues 메서드
추가
POIService 클래스에서 global queue 관련 로직 삭제

donggeonoh commented:
@beggu84 
> POIInfo, POIMO, PoiInfoMO .. 헤깔리네요ㅎㅎ
MO 라고 줄여서 붙이는 거는 일반적인 컨벤션인가요? 아니면 취향인가요?

안녕하세요 늦은 밤까지 리뷰 남겨주셔서 정말 감사합니다 ㅎㅎ
setValues 문제는 해결했고, MO를 붙이는 이유는 [reference](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CoreData/KeyConcepts.html#//apple_ref/doc/uid/TP40001075-CH30-SW1)를 보면서 엔티티 네임과 클래스 네임이 일치 할 필요가 없다 하여 Objective-C 에서 권장하는 네임 패턴을 설명하는데 그대로 적용하였습니다.. POI 같은 경우 이미 VO에서 사용중이여서 MO를 붙이는 방향으로 하게 되었습니다

eunjeongS2 reviewed (approved):
LGTM~~~ 고생하셨습니다 ~~~

rnfxl92 reviewed (approved):
LGTM
